state/state.go

```diff
@@ -64,85 +72,224 @@ func NewLedgerState(dataDir string, eventBus *event.EventBus, logger *slog.Logge
 	ls.Lock()
 	defer ls.Unlock()
 	e := evt.Data.(ChainsyncEvent)
 	if e.Rollback {
-		// Remove rolled-back blocks in reverse order
-		var tmpBlocks []models.Block
-		result := ls.db.Metadata().Where("slot > ?", e.Point.Slot).Order("slot DESC").Find(&tmpBlocks)
-		if result.Error != nil {
+		if err := ls.handleEventChainSyncRollback(e); err != nil {
+			// TODO: actually handle this error
 			ls.logger.Error(
-				fmt.Sprintf("failed to query blocks from ledger: %s", result.Error),
+				fmt.Sprintf("failed to handle rollback: %s", err),
 			)
 			return
 		}
-		for _, tmpBlock := range tmpBlocks {
-			if err := ls.removeBlock(tmpBlock); err != nil {
-				ls.logger.Error(
-					fmt.Sprintf("failed to remove block: %s", err),
-				)
-				return
-			}
-		}
-		// Generate event
-		ls.eventBus.Publish(
-			ChainRollbackEventType,
-			event.NewEvent(
-				ChainRollbackEventType,
-				ChainRollbackEvent{
-					Point: e.Point,
-				},
-			),
-		)
-		ls.logger.Info(fmt.Sprintf(
-			"chain rolled back, new tip: %x at slot %d",
-			e.Point.Hash,
-			e.Point.Slot,
-		))
 	} else {
-		// Add block to database
-		tmpBlock := models.Block{
-			Slot: e.Point.Slot,
-			Hash: e.Point.Hash,
-			// TODO: figure out something for Byron. this won't work, since the
-			// block number isn't stored in the block itself
-			Number: e.Block.BlockNumber(),
-			Type:   e.Type,
-			Cbor:   e.Block.Cbor(),
-		}
-		if err := ls.AddBlock(tmpBlock); err != nil {
+		if err := ls.handleEventChainSyncBlock(e); err != nil {
+			// TODO: actually handle this error
 			ls.logger.Error(
-				fmt.Sprintf("failed to add block to ledger state: %s", err),
+				fmt.Sprintf("failed to handle block: %s", err),
 			)
 			return
 		}
-		// Generate event
-		ls.eventBus.Publish(
+	}
+}
+
+func (ls *LedgerState) handleEventChainSyncRollback(e ChainsyncEvent) error {
+	// Remove rolled-back blocks in reverse order
+	var tmpBlocks []models.Block
+	result := ls.db.Metadata().Where("slot > ?", e.Point.Slot).Order("slot DESC").Find(&tmpBlocks)
+	if result.Error != nil {
+		return fmt.Errorf("query blocks: %w", result.Error)
+	}
+	for _, tmpBlock := range tmpBlocks {
+		if err := ls.removeBlock(tmpBlock); err != nil {
+			return fmt.Errorf("remove block: %w", err)
+		}
+	}
+	// Delete rolled-back UTxOs
+	var tmpUtxos []models.Utxo
+	result = ls.db.Metadata().Where("added_slot > ?", e.Point.Slot).Order("id DESC").Find(&tmpUtxos)
+	if result.Error != nil {
+		return fmt.Errorf("remove rolled-backup UTxOs: %w", result.Error)
+	}
+	for _, utxo := range tmpUtxos {
+		if err := models.UtxoDelete(ls.db, utxo); err != nil {
+			return fmt.Errorf("remove rolled-back UTxO: %w", err)
+		}
+	}
+	// Restore spent UTxOs
+	result = ls.db.Metadata().Where("deleted_slot > ?", e.Point.Slot).Find(&tmpUtxos)
+	if result.Error != nil {
+		return fmt.Errorf("restore spent UTxOs after rollback: %w", result.Error)
+	}
+	for _, utxo := range tmpUtxos {
+		utxo.DeletedSlot = 0
```

Comment:
It's best to UPDATE them at once in SQL.

Reply:
I switched this one to a single UPDATE query